File: day12/day12.py
'''
Note: Doing it in python first to get the idea down since I haven't used BFS much yet
'''
class Point:
    def __init__(self, letter, row, col) -> None:
        self.letter = letter
        self.row = row
        self.col = col
        self.dist = None

        if letter == 'S':
            self.value = 1
            self.dist = 0
        elif letter == 'E':
            self.value = ord('z') - 96 
        else:
            self.value = ord(letter) - 96 

class Map:
    def __init__(self) -> None:
        self.grid = []

class Solution:
    def __init__(self) -> None:
        self.grid = []
        self.found = False
        self.q = []

f = open("input.txt", "r")
line_arr = [list(line[0:-1]) for line in f.readlines()]
f.close()

map, solution = Map(), Solution()

#generate map.grid and solution.grid
for row in range(len(line_arr)):
    grid_row, visited_row = [], []
    
    for col in range(len(line_arr[0])):
        p = Point(line_arr[row][col], row, col)
        if line_arr[row][col] == 'S':
            solution.start = [row, col]
            p.dist = 0
            visited_row.append(True)
        else:
            visited_row.append(False)

        grid_row.append(p)
 
    map.grid.append(grid_row)
    solution.grid.append(visited_row)

#add start S to q
solution.q.append(map.grid[solution.start[0]][solution.start[1]])

#Dijkstra BFS search
while solution.found == False:
    el = solution.q.pop(0)
    if el.letter == 'E':
        solution.found = True
        print(f"The distance is {el.dist}")
    else:
        # check if neighbor el (left, top, right, bottom) are viable and not visited, then add distance that has been walked up to that node and then add to q
        l = map.grid[el.row][el.col-1] if el.col-1 >= 0 else None
        t = map.grid[el.row-1][el.col] if el.row-1 >= 0 else None
        r = map.grid[el.row][el.col+1] if el.col+1 < len(map.grid[0]) else None
        b = map.grid[el.row+1][el.col] if el.row+1 < len(map.grid) else None

        if l and solution.grid[l.row][l.col] == False and l.value - el.value <= 1:
            if l.dist == None or l.dist > el.dist + 1:
                l.dist = el.dist + 1
            solution.q.append(l)
            solution.grid[l.row][l.col] = True

        if t and solution.grid[t.row][t.col] == False and t.value - el.value <= 1:
            if t.dist == None or t.dist > el.dist + 1:
                t.dist = el.dist + 1
            solution.q.append(t)
            solution.grid[t.row][t.col] = True

        if r and solution.grid[r.row][r.col] == False and r.value - el.value <= 1:
            if r.dist == None or r.dist > el.dist + 1:
                r.dist = el.dist + 1
            solution.q.append(r)
            solution.grid[r.row][r.col] = True

        if b and solution.grid[b.row][b.col] == False and b.value - el.value <= 1:
            if b.dist == None or b.dist > el.dist + 1:
                b.dist = el.dist + 1
            solution.q.append(b)
            solution.grid[b.row][b.col] = True

File: day12/day12.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
static int BUFFSIZE = 100;

typedef struct Point {
    char letter[1];
    int row;
    int col; 
    int dist; // -1 for None
    int value; 
    int visited; // -1 for false, 1 for True
} Point;

typedef struct q_el {
    Point* item;
    struct q_el* next;
} q_el;

void enqueue(q_el* head, Point* p) {
    if (head->item == NULL) {
        head->item = p;
    } else {
        q_el* new;
        new = malloc(sizeof(q_el));
        new->item = p;
        new->next = NULL;
        while (head->next != NULL) {
            head = head->next;
        }
        head->next = new;
    }
}

void dequeue(q_el* head) { 
    if (head->item == NULL) {
        return;
    } else if (head->next == NULL) {
        head->item = NULL;
    } else {
        head->item = head->next->item;
        head->next = head->next->next;
    }  
}

int main () {
    FILE* fp = fopen("input.txt", "r");
    Point* map[BUFFSIZE][BUFFSIZE];
    int start_row; int start_col; int row_counter = 0;

    for ( ;; ) {
        char line[120];
        fgets(line, 120, fp);
        if (feof( fp )) break;

        for (int i = 0; i <= strlen(line); i++) {
            if (line[i] == '\n') { 
                break; 
            } else {
                
                Point* p;
                p = (Point*)malloc(sizeof(Point));
                p->letter[0] = line[i];
                p->row = row_counter;
                p->col = i;
                p->dist = -1;
                p->visited = -1;

                if (line[i] == 'S') {
                    p->value = 1;
                    start_row = row_counter;
                    start_col = i;
                } else if (line[i] == 'E') {
                    p->value = 26;
                } else {
                    p->value = (int)(line[i]) - 96;
                }
                map[row_counter][i] = p;
            }
        }
        row_counter++;
    }
    fclose(fp);

    size_t size = sizeof(map[0]);
    q_el head = { .item = map[start_row][start_col], .next = NULL};
    head.item->visited = 1;
    head.item->dist = 0;

    while (head.item != NULL) {
        Point* current = head.item;
        dequeue(&head);
        if (current->letter[0] == 'E') {
            printf("Distance is %d \n", current->dist);
            break;
        } else {
            Point* point_arr[4]; 
            current->col - 1 >= 0   ? point_arr[0]  = map[current->row][current->col-1] : NULL; // left
            current->col + 1 < size ? point_arr[1]  = map[current->row][current->col+1] : NULL; // right
            current->row - 1 >= 0   ? point_arr[2]  = map[current->row-1][current->col] : NULL; // top
            current->row + 1 < size ? point_arr[3]  = map[current->row+1][current->col] : NULL; // bottom

            for (int i = 0; i < 4; i++) {
                if (point_arr[i] != NULL && point_arr[i]->visited != 1 && point_arr[i]->value - current->value <= 1) {
                    if (point_arr[i]->dist == -1) {
                        point_arr[i]->dist = current->dist +1;
                    }
                    point_arr[i]->visited = 1;
                    enqueue(&head, point_arr[i]);
                }
            }
        }
    }
}